client/state/imports/reducer.js

```diff
@@ -0,0 +1,290 @@
+ */
+const incrementCount = state => ( { ...state, count: state.count + 1 } );
+
+/**
+ * Merge an update into the api property of the store.
+ *
+ * @param {Object} apiUpdate The object to merge into the api object
+ * @returns {Object} The updated state with api changes
+ */
+const mergeApiUpdate = curry( ( apiUpdate, state ) => ( {
+	...state,
+	api: {
+		...state.api,
+		...apiUpdate,
+	},
+} ) );
+
+/**
+ * Add a new importer to the object containing imports. If an importer already
+ * exists with the same id, this will overwrite it.
+ *
+ * @param {string} importerId The id of the importer
+ * @param {Object} importer The importer
+ * @param {Object} state The current redux store state
+ * @returns {Object} The updated state with the new importer added
+ */
+const createImporter = curry( ( importerId, importer, state ) => ( {
+	...state,
+	importers: {
+		...state.importers,
+		[ importerId ]: importer,
+	},
+} ) );
+
+/**
+ * Merge an update into the importer with the given id.
+ *
+ * @param {string} importerId The id of the importer to update
+ * @param {Object} importerUpdate An object containing the properties to update
+ * @param {Object} state The current redux store state
+ * @returns {Object} The updated state with the merged update
+ */
+const mergeImporterUpdate = curry( ( importerId, importerUpdate, state ) => ( {
+	...state,
+	importers: {
+		...state.importers,
+		[ importerId ]: {
+			...state.importers[ importerId ],
+			...importerUpdate,
+		},
+	},
+} ) );
+
+/**
+ * Remove an importer with the provided id.
+ *
+ * @param {Object} importerId The id of the importer to remove
+ * @param {Object} state The current redux store state
+ * @returns {Object} The updated state without the removed importer
+ */
+const deleteImporter = curry( ( importerId, state ) => ( {
+	...state,
+	importers: omit( state.importers, importerId ),
+} ) );
+
+/**
+ * importerStates that can be removed by the `deleteRemovableImporters` function
+ *
+ * @type {Array}
+ */
+const removableStates = [ appStates.CANCEL_PENDING, appStates.DEFUNCT ];
+
+/**
+ * Remove any importers at the states within the `removeableStates` array.
+ *
+ * @param {Object} state The current redux store state
+ * @returns {Object} The updated state without the removed importers
+ */
+const deleteRemovableImporters = state => ( {
+	...state,
+	importers: omitBy( state.importers, importer =>
+		includes( removableStates, importer.importerState )
+	),
+} );
+
+/**
+ * Returns the updated items state after an action has been dispatched.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export default createReducer( initialState, {
+	[ IMPORTS_STORE_RESET ]: () => initialState,
+	[ IMPORTS_FETCH ]: mergeApiUpdate( {
+		isFetching: true,
+	} ),
+	[ IMPORTS_FETCH_FAILED ]: state =>
+		mergeApiUpdate(
+			{
+				isFetching: false,
+				retryCount: state.api.retryCount + 1,
+			},
+			state
+		),
```

Comment:
Can `state => mergeApiUpdate( _, state )` be replaced with `mergeApiUpdate( _ )`?